skt/__init__.py

```diff
@@ -518,24 +539,73 @@ def mktgz(self, clean=True):
-                break
+        with io.open(self.buildlog, 'wb') as writer, \
+                io.open(self.buildlog, 'rb') as reader:
+            make = subprocess.Popen(args,
+                                    stdout=writer,
+                                    stderr=subprocess.STDOUT)
+            make.timedout = False
+
+            def stop_process(proc):
+                """
+                Terminate the process with SIGTERM and flag it as timed out.
+                """
+                if proc.poll() is None:
+                    proc.terminate()
+                    proc.timedout = True
+            timer = Timer(timeout, stop_process, [make])
+            timer.setDaemon(True)
+            timer.start()
+            while make.poll() is None:
+                self.append_and_log2stdout(reader.readlines(), stdout_list)
+                time.sleep(1)
+            self.append_and_log2stdout(reader.readlines(), stdout_list)
+            timer.cancel()
+            if make.timedout:
+                raise CommandTimeoutError(
+                    "'{}' was taking too long".format(' '.join(args))
+                )
+            if make.returncode != 0:
+                raise subprocess.CalledProcessError(make.returncode,
+                                                    ' '.join(args))
+
+        match = re.search("^Tarball successfully created in (.*)$",
+                          ''.join(stdout_list), re.MULTILINE)
+        if match:
+            fpath = os.path.realpath(os.path.join(self.path, match.group(1)))
+        else:
+            raise RegexNotFound('Failed to find tgz path in stdout')
 
-        fpath = None
-        if tgzpath is not None:
-            fpath = "/".join([self.path, tgzpath])
+        if not os.path.isfile(fpath):
+            raise IOError("Built kernel tarball {} not found".format(fpath))
 
-        if fpath is None or not os.path.isfile(fpath):
-            with open(self.buildlog, "w") as fp:
-                fp.write(stdout)
+        return fpath
 
-            raise Exception("Failed to find tgz path in stdout")
+    @staticmethod
+    def append_and_log2stdout(lines, full_log):
+        """
+        Append `lines` into `full_log` and show `lines` on stdout.
 
-        return fpath
+        Args:
+            lines:      list of strings.
+            full_log:   list where `lines` members are appended.
+        """
+        full_log.extend(lines)
+        sys.stdout.write(''.join(lines))
+        sys.stdout.flush()
+
+
+class CommandTimeoutError(Exception):
+    """
+    Exception raised when a timeout occurs on a process which has had timeouts
+    enabled. The accompanying value is a string whose value is the command
+    launched plus a small explanation.
+    """
+
+
+class RegexNotFound(Exception):
+    """
+    Exception raised when a regex does not match and it is impossible to
+    continue. The accompanying value is a string which explains what it can not
+    find.
+    """
```

Comment:
Do you think you can name this something like `ParsingError`, or similar? The `RegexNotFound` is a bit too specific, while at the same time not that helpful.

Reply:
OK, I'll rename it to `ParsingError` :+1: